src/main/java/com/aws/greengrass/device/iot/IotAuthClient.java

```diff
@@ -125,5 +135,75 @@ public boolean isThingAttachedToCertificate(Thing thing, Certificate certificate
             }
         }
+
+        @Override
+        public void clearLocalAuthCache() {
+            certificateHashToIdMap.clear();
+            certificateIdToThingMap.clear();
+        }
+
+        @SuppressWarnings("PMD.AvoidCatchingGenericException")
+        private Optional<String> fetchActiveCertificateId(String certificatePem) {
+            VerifyClientDeviceIdentityRequest request =
+                    VerifyClientDeviceIdentityRequest.builder().clientDeviceCertificate(certificatePem).build();
+            try {
+                VerifyClientDeviceIdentityResponse response = RetryUtils.runWithRetry(SERVICE_EXCEPTION_RETRY_CONFIG,
+                        () -> clientFactory.getGreengrassV2DataClient().verifyClientDeviceIdentity(request),
+                        "verify-client-device-identity", logger);
+                return Optional.of(response.clientDeviceCertificateId());
+            } catch (InterruptedException e) {
+                logger.atError().cause(e).log("Verify client device identity got interrupted");
+                // interrupt the current thread so that higher-level interrupt handlers can take care of it
+                Thread.currentThread().interrupt();
+                throw new CloudServiceInteractionException(
+                        "Failed to verify client device identity, process got interrupted", e);
+            } catch (ValidationException | ResourceNotFoundException e) {
+                logger.atWarn().cause(e).kv("certificatePem", certificatePem)
+                        .log("Certificate doesn't exist or isn't active");
+                return Optional.empty();
+            } catch (Exception e) {
+                logger.atError().cause(e).kv("certificatePem", certificatePem)
+                        .log("Failed to verify client device identity with cloud. Check that the core device's IoT "
+                                + "policy grants the greengrass:VerifyClientDeviceIdentity permission.");
+                throw new CloudServiceInteractionException("Failed to verify client device identity", e);
+            }
+        }
+
+        private Optional<String> getCertificateIdAvailableLocally(String certificatePem) {
+            try {
+                String certHash = sha256Hash(certificatePem);
+                if (certificateHashToIdMap.containsKey(certHash)) {
+                    return Optional.of(certificateHashToIdMap.get(certHash));
+                }
+            } catch (NoSuchAlgorithmException e) {
+                return Optional.empty();
+            }
+            return Optional.empty();
+        }
+
+        private void attachCertificateIdToPem(String certificateId, String certificatePem) {
+            try {
+                certificateHashToIdMap.put(sha256Hash(certificatePem), certificateId);
+            } catch (NoSuchAlgorithmException e) {
+                logger.atTrace().cause(e).log("Could not store CertificatePem to CertificateId mapping");
```

Comment:
This is one of those "impossible" things that a customer would want to know about. Failure to save cert ID information means GG can't work offline. This should be at least warn. Maybe error

Reply:
log level set to `WARN`